File: 1477-find-two-non-overlapping-sub-arrays-each-with-target-sum/1477-find-two-non-overlapping-sub-arrays-each-with-target-sum.js
/**
 * @param {number[]} arr
 * @param {number} target
 * @return {number}
 */
var minSumOfLengths = function(arr, target) {
    const n = arr.length
    const map = {}
    let sum = 0
    let size = Infinity
    let res = Infinity
    map[0] = -1
    for (let i=0; i < n; i++) {
        sum += arr[i]
        map[sum] = i
    }
    sum = 0
    for (let i=0; i < n; i++) {
        sum += arr[i]
        if ((sum - target) in map) {
            size = Math.min(size, i - map[sum - target])
        }
        if ((sum + target) in map && size < Infinity) {
            res = Math.min(res, map[sum + target] - i + size)
        }
    }
    return res === Infinity ? -1 : res
};

minSumOfLengths = function(arr, target) {
    const n = arr.length
    const best = Array(n).fill(Infinity)
    let res = Infinity
    let bestSoFar = Infinity
    let sum = 0
    let start = 0
    for (let i=0; i < n; i++) {
        sum += arr[i]
        while (sum > target) {
            sum -= arr[start]
            start++
        }
        if (sum === target) {
            if (start > 0 && best[start - 1] !== Infinity) {
                res = Math.min(res, best[start - 1] + i - start + 1)
            }
            bestSoFar = Math.min(bestSoFar, i - start + 1)
        }
        best[i] = bestSoFar
    }
    return res === Infinity ? -1 : res
}

File: 1477-find-two-non-overlapping-sub-arrays-each-with-target-sum/1477-find-two-non-overlapping-sub-arrays-each-with-target-sum.py
class Solution:
    def minSumOfLengths(self, arr: List[int], target: int) -> int:
        
        n = len(arr)
        best = [inf] * n
        best_so_far = inf
        res = inf
        cur_sum = 0
        start = 0
        
        for i in range(n):
            cur_sum += arr[i]
            
            while cur_sum > target:
                cur_sum -= arr[start]
                start += 1
            
            if cur_sum == target:
                if start > 0 and best[start-1] != inf:
                    left_arr_len = best[start - 1]
                    right_arr_len = i - start + 1
                    res = min(res, left_arr_len + right_arr_len)
                    
                best_so_far = min(best_so_far, i - start + 1)
                
            best[i] = best_so_far
                
        return res if res != inf else -1